src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -65,12 +65,48 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
-            using (var conversation = new SaslConversation(description.ConnectionId))
+            var firstStepResult = description.IsMasterResult.SpeculativeAuthenticate;
+
+            if (firstStepResult == null)
+            {
+                Authenticate(connection, description, previousStep: null, previousStepResult: null, cancellationToken);
+            }
+            else
+            {
+                var (firstStep, _) = CreateFirstStepAndCommand(_mechanism, connection, conversation: null, description);
+                Authenticate(
+                    connection,
+                    description,
+                    previousStep: firstStep,
+                    previousStepResult: firstStepResult,
+                    cancellationToken);
+            }
+        }
+
+        private void Authenticate(
+            IConnection connection,
+            ConnectionDescription description,
+            ISaslStep previousStep,
+            BsonDocument previousStepResult,
+            CancellationToken cancellationToken)
+        {
+            Ensure.IsNotNull(connection, nameof(connection));
+            Ensure.IsNotNull(description, nameof(description));
+            if (previousStep != null)
             {
-                var currentStep = _mechanism.Initialize( connection, conversation, description);
+                Ensure.IsNotNull(previousStepResult, nameof(previousStepResult));
+            }
 
-                var command = CreateStartCommand(currentStep);
-                while (true)
+            using (var conversation = new SaslConversation(description.ConnectionId))
+            {
+                /* We cannot combine the two statements below into a single statement without adding a reference to
+                 * ValueTuple */
+                var currentStepAndCommand = previousStep == null
+                    ? CreateFirstStepAndCommand(_mechanism, connection, conversation, description)
+                    : CreateNextStepAndCommand(conversation, previousStep, previousStepResult);
+                var (currentStep, command) = currentStepAndCommand;
```

Comment:
We're able to use this syntax thanks to the `Deconstruct` method here:
https://github.com/vincentkam/mongo-csharp-driver/pull/95/files#diff-7cb6644a6c07129601487f26bb0f9b42R480

Reply:
For some reason, combining the two statements above yields the following error:
```
  SaslAuthenticator.cs(167, 23): [CS8128] Member '.ctor' was not found on type '(, )' from assembly 'MongoDB.Driver.Core, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
```

This was confusing to me, so I added the comment to clarify why this happens

Reply:
I was able to get around the need for two statements by adding a hard-cast that always succeeds. I've updated the code and comments to to reflect this